core/windows_and_tabs/window_layout.py

```diff
@@ -0,0 +1,265 @@
+import copy
+from dataclasses import dataclass
+from time import perf_counter
+from typing import List, Optional, Union
+
+from talon import Context, Module, actions, settings, ui
+
+"""Tools for laying out windows in an arrangement """
+
+from talon import Context, Module, actions, settings
+from talon.ui import UIErr, Window
+
+mod = Module()
+
+mod.list(
+    "window_split_positions",
+    "Predefined window positions when splitting the screen between multiple windows.",
+)
+
+
+def focus_callback(_):
+    global layout_in_progress
+    global last_layout
+    if layout_in_progress is not None:
+        return
+
+    delta = 1
+    if last_layout is not None:
+        delta = perf_counter() - last_layout.finish_time
+    if delta >= 1:
+        last_layout = None
+
+
+@dataclass
+class WindowLayout:
+    """Represents a layout of windows on a screen"""
+
+    layout: list[str]
+    windows: list[Window]
+    should_rotate: bool
+    finish_time: float
+
+
+layout_in_progress: Optional[WindowLayout] = None
+last_layout: Optional[WindowLayout] = None
+
+
+SPLIT_POSITIONS = {
+    "split": {
+        2: ["Left", "Right"],
+        3: [
+            "LeftThird",
+            "CenterThird",
+            "RightThird",
+        ],
+    },
+    "clock": {
+        3: [
+            "Left",
+            "TopRight",
+            "BottomRight",
+        ],
+    },
+    "counterclock": {
+        3: [
+            "Right",
+            "TopLeft",
+            "BottomLeft",
+        ],
+    },
+}
+
+
+def snap_next(windows: list[Window], target_layout: str) -> Optional[int]:
```

Comment:
Add a dock string explaining what this function does and what it does especially when it returns none at the end